collie/model/hybrid_matrix_factorization.py

```diff
@@ -264,29 +362,62 @@ def _setup_model(self, **kwargs) -> None:
-            metadata_layers_dims = (
-                [self.hparams.item_metadata_num_cols] + self.hparams.metadata_layers_dims
+        # set up item metadata-only layers
+        item_metadata_output_dim = self.hparams.item_metadata_num_cols
+        self.item_metadata_layers = None
+
+        if self.hparams.item_metadata_layers_dims is not None:
+            item_metadata_layers_dims = (
+                [self.hparams.item_metadata_num_cols] + self.hparams.item_metadata_layers_dims
             )
-            self.metadata_layers = [
-                nn.Linear(metadata_layers_dims[idx - 1], metadata_layers_dims[idx])
-                for idx in range(1, len(metadata_layers_dims))
+            self.item_metadata_layers = [
+                nn.Linear(item_metadata_layers_dims[idx - 1], item_metadata_layers_dims[idx])
+                for idx in range(1, len(item_metadata_layers_dims))
             ]
-            for i, layer in enumerate(self.metadata_layers):
-                nn.init.xavier_normal_(self.metadata_layers[i].weight)
-                self.add_module('metadata_layer_{}'.format(i), layer)
-
-            metadata_output_dim = metadata_layers_dims[-1]
-
-        # set up combined layers
-        combined_dimension_input = (
-            self.user_embeddings.embedding_dim
-            + self.item_embeddings.embedding_dim
-            + metadata_output_dim
-        )
+            for i, layer in enumerate(self.item_metadata_layers):
+                nn.init.xavier_normal_(self.item_metadata_layers[i].weight)
+                self.add_module('item_metadata_layer_{}'.format(i), layer)
+
+            item_metadata_output_dim = item_metadata_layers_dims[-1]
+
+        # set up user metadata-only layers
+        user_metadata_output_dim = self.hparams.user_metadata_num_cols
+        self.user_metadata_layers = None
+
+        if self.hparams.user_metadata_layers_dims is not None:
+            user_metadata_layers_dims = (
+                [self.hparams.user_metadata_num_cols] + self.hparams.user_metadata_layers_dims
+            )
+            self.user_metadata_layers = [
+                nn.Linear(user_metadata_layers_dims[idx - 1], user_metadata_layers_dims[idx])
+                for idx in range(1, len(user_metadata_layers_dims))
+            ]
+            for i, layer in enumerate(self.user_metadata_layers):
+                nn.init.xavier_normal_(self.user_metadata_layers[i].weight)
+                self.add_module('user_metadata_layer_{}'.format(i), layer)
+
+            user_metadata_output_dim = user_metadata_layers_dims[-1]
+
+        # set up combined layers depending on metadata inputs
+        if item_metadata_output_dim is not None and user_metadata_output_dim is not None:
+            combined_dimension_input = (
+                self.user_embeddings.embedding_dim
+                + self.item_embeddings.embedding_dim
+                + item_metadata_output_dim
+                + user_metadata_output_dim
```

Comment:
I can't find the paper now for the life of me, but I read a paper where they suggested putting relevant metadata close to the embedding the metadata references. I have no empirical evidence that it would work, but it would make this code block turn into: 

```suggestion
                user_metadata_output_dim
                + self.user_embeddings.embedding_dim
                + self.item_embeddings.embedding_dim
                + item_metadata_output_dim 
```

Do you think that will actually matter to the model? 

Same thing for the user metadata-only situation in L415.

Reply:
even if it doesn't work it does give the code a certain symmetry, which is nice 😄